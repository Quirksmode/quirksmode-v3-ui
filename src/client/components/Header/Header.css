@import './LogoNav/LogoNav';

.Header {
  background: var(--color-primary-red);
  background-size: 100px 100px;
  border-bottom: 7px solid var(--color-grey-bg);
  position: relative;
  z-index: var(--z-index-Header);

  &__innerWrap {
    padding: 46px 0;

    @media (--mobile-landscape) {
      padding-bottom: 0;
    }

    @media (--tablet-portrait) {
      padding: 70px 0 0;
    }
  }

  &__sliderOuterWrap {
    @media (--mobile-landscape) {
      padding: 0 40px;
    }

    @media (--tablet-portrait) {
      padding: 0;
    }
  }

  &__sliderWrap {
    background: transparent url(../img/sliderShell-tablet.png) 0 0 no-repeat;
    background-size: contain;
    bottom: -34px;
    display: none;
    padding: 7.4% 9.5% 10% 10.1%;
    position: relative;
    z-index: var(--z-index-Header__sliderWrap);

    @include retina-bg(../img/sliderShell-tablet, png);

    @media (--mobile-landscape) {
      display: block;
    }

    @media (--tablet-portrait) {
      @include retina-bg(../img/sliderShell-mac, png, min-width: 768px);
      background: transparent url(../img/sliderShell-mac.png) 0 0 no-repeat;
      background-size: contain;
      padding: 3.4% 11.9% 9% 12.1%;
    }

    &::after {
      background: url(../img/reflection-mac.png) 96.3% 4px no-repeat;
      background-size: contain;
      content: '';
      height: 88.5%;
      pointer-events: none;
      position: absolute;
      right: 0;
      top: 0;
      width: 100%;
      z-index: var(--z-index-Header__sliderWrap__after);

      @include retina-bg(../img/reflection-mac, png);

      @media (--tablet-portrait) {
        background-position: 86.2% 4px;
        height: 88%;
      }
    }
  }

  .swiper-wrapper {
    height: auto;
    padding-bottom: 5.5%;
  }

  .swiper-container {
    margin-bottom: -5.5%;
  }

  .swiper-pagination {
    bottom: 0;

    @media (--tablet-portrait) {
      bottom: 4.1%;
    }

    @media (--tablet-landscape) {
      bottom: 22px;
    }
  }

  .slider {
    &__directionNav {
      display: none;
      opacity: 0;

      &--active {
        opacity: 1;
      }

      @media (--mobile-landscape) {
        display: block;
        top: 65.4%;
      }

      @media (--tablet-portrait) {
        top: 60.5%;
      }
    }
  }
}
