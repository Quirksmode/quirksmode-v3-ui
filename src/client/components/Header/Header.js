import React, {
  useState
} from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Swiper from 'react-id-swiper';
import { useInView } from 'react-intersection-observer';
import LazyLoad from 'components/LazyLoad/LazyLoad';
import LogoNav from './LogoNav/LogoNav';

/**
 * Description
 *
 * @name Header
 * @param  {object}   props.app                    []
 * @param  {object}   props.mainLogo               []
 * @param  {array}    props.heroSlides             []
 * @param  {object}   props.location               []
 */
const Header = ({
  pageHomeContent,
  location
}) => {
  const {
    heroSlides
  } = pageHomeContent;
  /**
   *
   * @name swiper
   * @type {object}
   */
  const [swiper, updateSwiper] = useState(null);

  /**
   *
   * @name params
   * @type {object}
   */
  const params = {
    autoHeight: true,
    preloadImages: false,
    loop: false,
    shouldSwiperUpdate: true,
    updateOnImagesReady: true,
    pagination: {
      el: '.swiper-pagination',
      clickable: true
    }
  };

  /**
   *
   * @name goNext
   * @type {function}
   */
  const goNext = () => {
    if (swiper !== null) {
      swiper.slideNext();
    }
  };

  /**
   *
   * @name goPrev
   * @type {function}
   */
  const goPrev = () => {
    if (swiper !== null) {
      swiper.slidePrev();
    }
  };

  /**
   * @description Hook that is generated by the useInView function from the
   *              react-intersection-observer library.
   *
   */
  const [sliderNavRef, inView] = useInView({
    threshold: 1,
    rootMargin: '-150px'
  });

  return (
    <header className="Header">
      <div className="Header__innerWrap grid">
        <LogoNav />
        { location.pathname === '/' && heroSlides && heroSlides.length > 0 && (
          <div className="Header__sliderOuterWrap">
            <div className="Header__sliderWrap">
              <Swiper
                getSwiper={ updateSwiper }
                { ...params }
              >
                { heroSlides.map(slide => (
                  <div className="aspectWrap aspectWrap--ratio-16-10" key={ slide.id }>
                    <LazyLoad>
                      <picture>
                        <source
                          type="image/webp"
                          srcSet={
                            `${slide.image.sizes.heroSlider768up}.webp 1x,
                            ${slide.image.sizes.heroSlider768up2x}.webp 1.5x,
                            ${slide.image.sizes.heroSlider768up2x}.webp 2x`
                          }
                          media="(min-width: 768px)"
                        />
                        <source
                          type="image/jpeg"
                          srcSet={
                            `${slide.image.sizes.heroSlider768up} 1x,
                            ${slide.image.sizes.heroSlider768up2x} 1.5x,
                            ${slide.image.sizes.heroSlider768up2x} 2x`
                          }
                          media="(min-width: 768px)"
                        />
                        <source
                          type="image/webp"
                          srcSet={
                            `${slide.image.sizes.heroSlider481up}.webp 1x,
                            ${slide.image.sizes.heroSlider481up2x}.webp 1.5x,
                            ${slide.image.sizes.heroSlider481up2x}.webp 2x`
                          }
                        />
                        <img
                          srcSet={
                            `${slide.image.sizes.heroSlider481up} 1x,
                            ${slide.image.sizes.heroSlider481up2x} 1.5x,
                            ${slide.image.sizes.heroSlider481up2x} 2x`
                          }
                          src={ slide.image.sizes.heroSlider481up }
                          alt={ slide.image.alt }
                          width={ slide.image.width }
                          height={ slide.image.height }
                        />
                      </picture>
                    </LazyLoad>
                  </div>
                ))}
              </Swiper>
            </div>
          </div>
        )}
      </div>

      { location.pathname === '/' && (
        <div
          ref={ sliderNavRef }
          className={ `Slider__directionNav${inView ? ' Slider__directionNav--active' : ''}` }
        >
          <button
            type="button"
            onClick={ goPrev }
            className="Slider__prev"
          >
            Previous
          </button>
          <button
            type="button"
            onClick={ goNext }
            className="Slider__next"
          >
            Next
          </button>
        </div>
      )}
    </header>
  );
};

Header.propTypes = {
  pageHomeContent: PropTypes.object,
  location: PropTypes.object,
};

const mapStateToProps = ({ pageHome }) => ({
  pageHomeContent: pageHome.content
});

export default connect(
  mapStateToProps,
  null
)(Header);
