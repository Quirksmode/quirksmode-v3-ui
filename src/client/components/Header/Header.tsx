import React, {
  useState
} from 'react';
import {
  useSelector
} from 'react-redux';
import PropTypes from 'prop-types';
import Swiper from 'react-id-swiper';
import { useInView } from 'react-intersection-observer';
import LazyLoad from 'components/LazyLoad/LazyLoad';
import LogoNav from './LogoNav/LogoNav';

/**
 * Description
 *
 * @name Header
 * @param  {object}   props.app                    []
 * @param  {object}   props.mainLogo               []
 * @param  {array}    props.heroSlides             []
 * @param  {object}   props.location               []
 */
const Header = ({
  location
}) => {
  const heroSlides = useSelector(state => state.pageHome.content.heroSlides);

  /**
   *
   * @name swiper
   * @type {object}
   */
  const [swiper, updateSwiper] = useState(null);

  /**
   *
   * @name params
   * @type {object}
   */
  const params = {
    autoplay: {
      delay: 7000
    },
    speed: 1000,
    fadeEffect: {
      crossFade: true
    },
    effect: 'fade',
    autoHeight: true,
    preloadImages: false,
    loop: false,
    shouldSwiperUpdate: true,
    updateOnImagesReady: true,
    pagination: {
      el: '.swiper-pagination',
      clickable: true
    }
  };

  /**
   *
   * @name goNext
   * @type {function}
   */
  const goNext = () => {
    if (swiper !== null) {
      if (swiper.activeIndex === (swiper.slides.length - 1)) {
        // Fake Loop due to React issue with Loop mode
        swiper.slideTo(0);
      } else {
        swiper.slideNext();
      }
    }
  };

  /**
   *
   * @name goPrev
   * @type {function}
   */
  const goPrev = () => {
    if (swiper !== null) {
      if (swiper.activeIndex === 0) {
        // Fake Loop due to React issue with Loop mode
        swiper.slideTo(swiper.slides.length - 1);
      } else {
        swiper.slidePrev();
      }
    }
  };

  /**
   * @description Hook that is generated by the useInView function from the
   *              react-intersection-observer library.
   *
   */
  const [sliderNavRef, inView] = useInView({
    threshold: 1,
    rootMargin: '-150px'
  });

  return (
    <header className="Header">
      <div className="Header__innerWrap grid">
        <LogoNav />
        { location.pathname === '/' && heroSlides && heroSlides.length > 0 && (
          <div className="Header__sliderOuterWrap">
            <div className="Header__sliderWrap">
              <Swiper
                data-test="Header__slider"
                getSwiper={ updateSwiper }
                { ...params }
              >
                { heroSlides.map(slide => (
                  <div className="aspectWrap aspectWrap--ratio-16-10" key={ slide.id }>
                    <LazyLoad>
                      <picture>
                        <source
                          type="image/webp"
                          srcSet={
                            `${slide.image.sizes.heroSlider768up2x}.webp 1x,
                            ${slide.image.sizes.heroSlider768up2x}.webp 1.5x,
                            ${slide.image.sizes.heroSlider768up2x}.webp 2x`
                          }
                          media="(min-width: 768px)"
                        />
                        <source
                          type="image/jpeg"
                          srcSet={
                            `${slide.image.sizes.heroSlider768up2x} 1x,
                            ${slide.image.sizes.heroSlider768up2x} 1.5x,
                            ${slide.image.sizes.heroSlider768up2x} 2x`
                          }
                          media="(min-width: 768px)"
                        />
                        <source
                          type="image/webp"
                          srcSet={
                            `${slide.image.sizes.heroSlider481up}.webp 1x,
                            ${slide.image.sizes.heroSlider481up2x}.webp 1.5x,
                            ${slide.image.sizes.heroSlider481up2x}.webp 2x`
                          }
                        />
                        <img
                          srcSet={
                            `${slide.image.sizes.heroSlider481up} 1x,
                            ${slide.image.sizes.heroSlider481up2x} 1.5x,
                            ${slide.image.sizes.heroSlider481up2x} 2x`
                          }
                          src={ slide.image.sizes.heroSlider481up }
                          alt={ slide.image.alt }
                          width={ slide.image.width }
                          height={ slide.image.height }
                        />
                      </picture>
                    </LazyLoad>
                  </div>
                ))}
              </Swiper>
            </div>
          </div>
        )}
      </div>

      { location.pathname === '/' && (
        <div
          ref={ sliderNavRef }
          className={ `Slider__directionNav${inView ? ' Slider__directionNav--active' : ''}` }
        >
          <button
            type="button"
            onClick={ goPrev }
            className="Slider__prev"
          >
            Previous
          </button>
          <button
            type="button"
            onClick={ goNext }
            className="Slider__next"
          >
            Next
          </button>
        </div>
      )}
    </header>
  );
};

Header.propTypes = {
  location: PropTypes.object
};

export default Header;
