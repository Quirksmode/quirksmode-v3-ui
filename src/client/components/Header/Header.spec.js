import {
  findByTestAttr,
  makeMountRender,
  reduxify,
  snapshotify
} from 'tests/testUtils';
import appStub from 'client/App.stub.json';
import pageHomeStub from 'pages/PageHome/PageHome.stub.json';
import { Header } from './Header';
import 'tests/mocks/intersectionObserver';

const props = {
  location: {
    pathname: '/'
  }
};

describe('Header', () => {
  it('matches snapshot', () => {
    const wrapper = makeMountRender(reduxify(Header, props))();
    expect(snapshotify(wrapper)).toMatchSnapshot();
  });

  describe('Render with initial props', () => {
    it('logo should not display', () => {
      const wrapper = makeMountRender(reduxify(Header, props))();
      const logo = findByTestAttr(wrapper, 'logo');
      expect(logo.length).toBe(0);
    });
  });

  describe('Render with fetched props', () => {
    it('logo should display', () => {
      const wrapper = makeMountRender(reduxify(Header, props, {
        app: appStub,
        pageHome: pageHomeStub
      }))();
      const logo = findByTestAttr(wrapper, 'logo');
      expect(logo.length).toBe(1);
    });
  });
});
