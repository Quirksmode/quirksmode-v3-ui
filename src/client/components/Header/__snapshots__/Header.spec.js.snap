// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header matches snapshot 1`] = `"<header class=\\"Header\\"><div class=\\"Header__innerWrap grid\\"><div class=\\"LogoNav\\"><nav class=\\"LogoNav__nav\\"><p class=\\"visuallyHidden\\"><strong>Main Navigation</strong></p><ul class=\\"LogoNav__nav__items\\"></ul></nav></div></div><div class=\\"Slider__directionNav\\"><button type=\\"button\\" class=\\"Slider__prev\\">Previous</button><button type=\\"button\\" class=\\"Slider__next\\">Next</button></div></header>"`;
